### ISSUE TITLE
Pool slower than serial, pickling with every call?

### ISSUE THREAD

**Issue by guillochon** [dfm/emcee#199]
Hi all, I'm seeing some pretty abysmal performance out of emcee in parallel, the code is even slower than serial on a single computer. I have a feeling this is because my objective function is buried in a class, and emcee pickles the entire class when it maps to the pool. Unfortunately, I don't believe it's easy to separate the objective function from the class, the function calls a variety of methods within the class and therefore the class needs to be distributed to the workers. However, I don't believe the class needs to be distributed *with every call*, which emcee currently seems to be doing.

Is there a way to "pre-distribute" the class without having to pass a large set of pickles with every call?

Part of my code where sampler is called is here: https://github.com/guillochon/MOSFiT/blob/master/mosfit/model.py#L359

**Comment by dfm** [comment]
Because of the way multiprocessing works, there's nothing that you can do about this without using a more sophisticated pool and workflow. I recommend using the MPIPool from [schwimmbad](https://github.com/adrn/schwimmbad) something like the following (Disclaimer: not directly tested!):

```
import emcee
from schwimmbad import MPIPool

# expensive steps of loading model, etc. - this will be run on every cpu
model = ...

def lnprob(params):
    return model.lnprob(params)

with MPIPool() as pool:
    if not pool.is_master():
        pool.wait()
        sys.exit(0)

    sampler = emcee.EnsembleSampler(..., pool=pool)
    sampler.run_mcmc(...)
```

**Comment by guillochon** [comment]
OK, I've managed to fix this with a wrapper constructed in a way similar to what you suggested. Thanks for your help! I have another issue that has cropped though, but I'll close this once since it's a bit different.

**Comment by gnarayan** [comment]
Following Dan's example (running into weird slowdown moving from emcee.utils.MPIPool to schwimmbad) can't one replace 
`
# expensive steps of loading model, etc. - this will be run on every cpu
model = ...
`

with 

`if pool.is_master():
    # expensive steps on one cpu
    model = ...
    pool.bcast(model)
` 

Asking because expensive steps include writing to disk, and I'd rather now have N processes trying to  write out the same file.

**Comment by dfm** [comment]
That should be fine too, I imagine, but I've never actually used it!
